Turbocharged formula name completion

metacollin opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----

This is a reworked formula name completion for the brew command that produces a name list that is identical to `$(brew search)` while being roughly 20-30 times faster, on average.  Note, the test below was done with 9 taps, some of them with dozens of Formulae, just to show that it is fast even with a lot of taps.

```bash
=> echo && echo "Old brew completion time:" && time brew_old_completion \
&& echo && echo "New brew completion time:" && time brew_new_completion

Old brew completion time:

real	0m4.114s
user	0m1.856s
sys	0m2.267s

New brew completion time:

real	0m0.134s
user	0m0.081s
sys	0m0.118s
```

It uses the high speed of find, sed, and grep to obtain and output that is functionally identical, but by traversing the Tap directory structure directly and avoids invoking `brew search`.  `brew search` is much more thorough, but in the special case of being invoked with no pattern, it falls back to a simpler output intended for use with completions,  but still takes the same time to run.  It seems like there is no advantage here to using brew's search command for completions when it can be done just as well using only default system binaries and switches that work just as readily with the BSD and GNU variants of those binaries.  I also derive the location of the file structure the same way the completion script is already doing it for completing tap names, so I think doing it for Formula as well should be ok.

I know it is not exactly the most readable solution, but regular expressions rarely are.  It isn't doing anything particularly exotic.  It uses find to traverse the tap directories, using -prune to ignore the Cask, cmd, and .git/hub directories which can contain non-formula .rb files, then produces a list of .rb files which are definitely formulas.  Sed removes the extra stuff we don't want (homebrew- prefixes, the full path, .rb extensions).  Then short names are also produced, and finally the extended names for the homebrew core tap is reduced to short names only.  This results in an identical completion list as `brew search` produces now.  

This method is very similar to how the brew cask name completion is already handled, which also does it at the filesystem level.  This just has a little bit more sed and uses find instead of ls as it needs to traverse multiple taps with inconsistent directory structures.  

If no one takes issue or otherwise finds a problem with doing it this way, I plan on porting it to fish, and possibly zsh, though zsh seems to cache the completion results so the speed up would have a minimal impact (it would mostly just speed up the first time formula name completion was used in that shell).

jonchang commented:
The speedup is really impressive, but since you use gfind, which doesn't ship by default on macOS or Linux, won't this cause errors for users who don't have `findutils` installed?

metacollin commented:
Whoops, that was a mistake.  In my own personal version, I use gfind because its a little faster, I must have committed it by accident.  It works with the BSD find as well, but takes about 30ms longer.   

Anyway, fixed, thanks for catching that.  I also added caching of the `HOMEBREW_REPO` variable which trims off another ~50ms.

MikeMcQuaid commented on completions/bash/brew:
```diff
@@ -50,9 +50,13 @@ __brewcomp() {
-  COMPREPLY=($(compgen -W "$formulas $shortnames" -- "$cur"))
+  HOMEBREW_REPOSITORY="${HOMEBREW_REPOSITORY:-$(brew --repo)}"
+
+  local formulae="$(\find $HOMEBREW_REPOSITORY/Library/Taps -type d \( -name Casks -o -name cmd -o -name .github \) -prune -false -o -name '*rb' | \sed 's/\.rb//g' | \sed -E 's/.*\/Taps\/(.*)\/homebrew-/\1\//g' | \sed 's/\/Formula\//\//g')"
```
This is great. I think it'd be better handled with a brew command implemented in Bash (e.g. `brew formulae`) that does the same thing (and can be used by e.g. ZSH completions too). How does that sound to you?

metacollin replied:
Sorry for taking so long to reply.  That sounds like a great idea. I'll take a look.

metacollin replied:
Hi, after playing around a bit, it looks like adding it as a proper command (under the cmd directory) also incurs a pretty hefty startup time cost and increases the run time to about ~700ms.

What do you think of doing this?

```diff
diff --git a/Library/Homebrew/brew.sh b/Library/Homebrew/brew.sh
index 814c0ca22..cc615162c 100644
--- a/Library/Homebrew/brew.sh
+++ b/Library/Homebrew/brew.sh
@@ -48,11 +48,15 @@ else
   HOMEBREW_CELLAR="$HOMEBREW_PREFIX/Cellar"
 fi
 
+# shellcheck disable=SC1090
+source "$HOMEBREW_LIBRARY/Homebrew/cmd/formulae.sh"
+
 case "$*" in
-  --prefix)            echo "$HOMEBREW_PREFIX"; exit 0 ;;
-  --cellar)            echo "$HOMEBREW_CELLAR"; exit 0 ;;
-  --repository|--repo) echo "$HOMEBREW_REPOSITORY"; exit 0 ;;
-  --caskroom)          echo "$HOMEBREW_PREFIX/Caskroom"; exit 0 ;;
+  --prefix)               echo "$HOMEBREW_PREFIX"; exit 0 ;;
+  --cellar)               echo "$HOMEBREW_CELLAR"; exit 0 ;;
+  --repository|--repo)    echo "$HOMEBREW_REPOSITORY"; exit 0 ;;
+  --caskroom)             echo "$HOMEBREW_PREFIX/Caskroom"; exit 0 ;;
+  --formulae|--formulas)  homebrew-formulae; exit 0 ;;
 esac
 
 # A depth of 1 means this command was directly invoked by a user.

```

This recovers almost all of the speed.  A --formulae_spec.rb test could even be written that would compare the output to brew search this way.  

I guess it depends on how important that 400ms is to everyone.  This makes the command feel subjectively a lot slower to me, but that's just me.

MikeMcQuaid replied:
> Hi, after playing around a bit, it looks like adding it as a proper command (under the cmd directory) also incurs a pretty hefty startup time cost and increases the run time to about ~700ms.

Is that with a hot or cold cache? If I repeatedly run `brew shellenv` about 0.5s is removed from the time on the second time it is run.

> What do you think of doing this?

👎🏻 on that, it's a bit of a hack. At most I could see an argument for having a `source`d shared script which is sourced directly in the ZSH completion.

MikeMcQuaid replied:
@metacolin Actually, you've convinced me 😂. I'm opening a PR to do this for `shellenv` and then you can do the same for `formulae`.

MikeMcQuaid replied:
https://github.com/Homebrew/brew/pull/8904. Once that's merged you can rebase this PR on there (or merge `master` in) and use the same format I've done there to add a new `formulae` command.

maxim-belkin replied:
@metacollin, for your `sed` commands -- you can use any other separators instead of `/` so that you don't have to escape `/` in patterns. Also, you can combine all three `sed`s into one:

```
\sed -E 's#\.rb##g; s#.*/Taps/(.*)/homebrew-#\1/#g; s#/Formula/#/#g'
```

Then, currently, this returns things like:
<details>

```
homebrew/portable-ruby/Abstract/portable-formula
homebrew/test-bot/spec/spec_helper
homebrew/test-bot/spec/homebrew/step_spec
homebrew/test-bot/spec/homebrew/tests/setup_spec
homebrew/test-bot/spec/stub/formula_installer
homebrew/test-bot/spec/stub/utils/bottles
homebrew/test-bot/spec/stub/tap
homebrew/test-bot/spec/stub/utils
homebrew/test-bot/spec/stub/os
homebrew/test-bot/spec/stub/development_tools
homebrew/test-bot/spec/stub/formula
homebrew/test-bot/spec/global
homebrew/test-bot/lib/test_bot
homebrew/test-bot/lib/tests/cleanup_before
homebrew/test-bot/lib/tests/cleanup_after
homebrew/test-bot/lib/tests/setup
homebrew/test-bot/lib/tests/tap_syntax
homebrew/test-bot/lib/tests/formulae
homebrew/test-bot/lib/step
homebrew/test-bot/lib/test_runner
homebrew/test-bot/lib/test
homebrew/test-bot/lib/test_cleanup
homebrew/test-bot/lib/test_ci_upload
homebrew/bundle/spec/spec_helper
homebrew/bundle/spec/bundle/cask_installer_spec
homebrew/bundle/spec/bundle/skipper_spec
homebrew/bundle/spec/bundle/brewfile_spec
homebrew/bundle/spec/bundle/cask_dumper_spec
homebrew/bundle/spec/bundle/mac_app_store_installer_spec
homebrew/bundle/spec/bundle/tap_dumper_spec
homebrew/bundle/spec/bundle/whalebrew_installer_spec
homebrew/bundle/spec/bundle/dumper_spec
homebrew/bundle/spec/bundle/dsl_spec
homebrew/bundle/spec/bundle/tap_installer_spec
homebrew/bundle/spec/bundle/brew_installer_spec
homebrew/bundle/spec/bundle/commands/check_command_spec
homebrew/bundle/spec/bundle/commands/exec_command_spec
homebrew/bundle/spec/bundle/commands/list_command_spec
homebrew/bundle/spec/bundle/commands/dump_command_spec
homebrew/bundle/spec/bundle/commands/install_command_spec
homebrew/bundle/spec/bundle/commands/cleanup_command_spec
homebrew/bundle/spec/bundle/brew_dumper_spec
homebrew/bundle/spec/bundle/mac_app_store_dumper_spec
homebrew/bundle/spec/bundle/locker_spec
homebrew/bundle/spec/bundle/brew_services_spec
homebrew/bundle/spec/bundle/whalebrew_dumper_spec
homebrew/bundle/spec/bundle_utils_spec
homebrew/bundle/spec/stub/extend/ENV
homebrew/bundle/spec/stub/exceptions
homebrew/bundle/spec/stub/utils/formatter
homebrew/bundle/spec/stub/tap
homebrew/bundle/spec/stub/utils
homebrew/bundle/spec/stub/os
homebrew/bundle/spec/stub/formulary
homebrew/bundle/spec/stub/development_tools
homebrew/bundle/spec/stub/global
homebrew/bundle/spec/stub/formula
homebrew/bundle/lib/bundle/extend/os/skipper
homebrew/bundle/lib/bundle/extend/os/linux/skipper
homebrew/bundle/lib/bundle/extend/os/linux/bundle
homebrew/bundle/lib/bundle/extend/os/bundle
homebrew/bundle/lib/bundle/checker
homebrew/bundle/lib/bundle/brew_installer
homebrew/bundle/lib/bundle/brew_services
homebrew/bundle/lib/bundle/lister
homebrew/bundle/lib/bundle/brew_service_checker
homebrew/bundle/lib/bundle/mac_app_store_dumper
homebrew/bundle/lib/bundle/brewfile
homebrew/bundle/lib/bundle/cask_installer
homebrew/bundle/lib/bundle/brew_checker
homebrew/bundle/lib/bundle/cask_dumper
homebrew/bundle/lib/bundle/installer
homebrew/bundle/lib/bundle/mac_app_store_installer
homebrew/bundle/lib/bundle/tap_dumper
homebrew/bundle/lib/bundle/whalebrew_dumper
homebrew/bundle/lib/bundle/dumper
homebrew/bundle/lib/bundle/cask_checker
homebrew/bundle/lib/bundle/skipper
homebrew/bundle/lib/bundle/tap_installer
homebrew/bundle/lib/bundle/locker
homebrew/bundle/lib/bundle/commands/install
homebrew/bundle/lib/bundle/commands/dump
homebrew/bundle/lib/bundle/commands/cleanup
homebrew/bundle/lib/bundle/commands/exec
homebrew/bundle/lib/bundle/commands/check
homebrew/bundle/lib/bundle/commands/list
homebrew/bundle/lib/bundle/tap_checker
homebrew/bundle/lib/bundle/bundle
homebrew/bundle/lib/bundle/whalebrew_installer
homebrew/bundle/lib/bundle/dsl
homebrew/bundle/lib/bundle/mac_app_store_checker
homebrew/bundle/lib/bundle/brew_dumper
homebrew/bundle/lib/bundle
homebrew/services/spec/spec_helper
homebrew/services/spec/homebrew/services_cli_spec
homebrew/services/spec/homebrew/service_spec
homebrew/services/spec/stub/exceptions
homebrew/services/lib/services_cli
homebrew/services/lib/service
```
</details>

metacollin replied:
Sorry, I got kind of swamped with other stuff.  Changes made.  @maxim-belkin Thanks for the separator tip, I changed the regexes accordingly.  However, using extended regular expressions is actually fairly expensive, so leaving the command split into 3 sed commands, but only one using the `-E` flag, is slightly faster (by like 10ms, you have to run a bunch of tests and average them to even measure it).  

@MikeMcQuaid I made the changes matching your style.  It certainly seems less hacky that way.  Thanks for the help, and sorry I took so long to get around to this.

metacollin replied:
Hmm, it looks like it is failing because brew man changed?  I'm not sure what to do, maybe someone needs to manually approve my documentation for the formulae.sh command?

jonchang commented:
You'll need to run `brew man` and then commit the changes.

metacollin commented:
Ah, whoops  Thanks!

> On Oct 24, 2020, at 7:00 PM, Jonathan Chang <notifications@github.com> wrote:
> 
> 
> You'll need to run brew man and then commit the changes.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Homebrew/brew/pull/8846#issuecomment-716076645>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AABMM6SFVOBQ2W2V5G2D743SMN2EHANCNFSM4SC2JMIQ>.
>

MikeMcQuaid commented on Library/Homebrew/brew.sh:
```diff
@@ -38,12 +38,13 @@ HOMEBREW_TEMP="${HOMEBREW_TEMP:-${HOMEBREW_DEFAULT_TEMP}}"
+  --caskroom)           echo "$HOMEBREW_PREFIX/Caskroom"; exit 0 ;;
+  --cache)              echo "$HOMEBREW_CACHE"; exit 0 ;;
+  shellenv)             source "$HOMEBREW_LIBRARY/Homebrew/cmd/shellenv.sh"; homebrew-shellenv; exit 0 ;;
+  formulae|formulas)    source "$HOMEBREW_LIBRARY/Homebrew/cmd/formulae.sh"; homebrew-formulae; exit 0 ;;
```
```suggestion
  formulae)    source "$HOMEBREW_LIBRARY/Homebrew/cmd/formulae.sh"; homebrew-formulae; exit 0 ;;
```

That's the naming we use consistently and I don't think we need to add an alias here.

MikeMcQuaid commented on Library/Homebrew/cmd/formulae.sh:
```diff
@@ -0,0 +1,11 @@
+  local formulae="$(find "$HOMEBREW_REPOSITORY"/Library/Taps -type d \( -name Casks -o -name cmd -o -name .github \) -prune -false -o -name '*rb' | sed 's/\.rb//g' | sed -E 's .*/Taps/(.*)/homebrew- \1/ g' | sed 's /Formula/ / g')"
+  local shortnames="$(echo "$formulae" | cut -d / -f 3)"
+  echo -e "$formulae \n $shortnames" | grep -v '^homebrew/' | sort -uf
+}
```
```suggestion
}

```

MikeMcQuaid commented on Library/Homebrew/cmd/formulae.sh:
```diff
@@ -0,0 +1,11 @@
+
+# shellcheck disable=SC2155
+homebrew-formulae() {
+  local formulae="$(find "$HOMEBREW_REPOSITORY"/Library/Taps -type d \( -name Casks -o -name cmd -o -name .github \) -prune -false -o -name '*rb' | sed 's/\.rb//g' | sed -E 's .*/Taps/(.*)/homebrew- \1/ g' | sed 's /Formula/ / g')"
```
```suggestion
  local formulae="$(
    find "$HOMEBREW_REPOSITORY"/Library/Taps -type d \( -name Casks -o -name cmd -o -name .github \) -prune -false -o -name '*\.rb' 
      | sed -E -e 's/\.rb//g' -e 's .*/Taps/(.*)/homebrew-\1/ g' -e 's /Formula/ / g')"
```

Some questions, though: 
- are you sure the `g` is needed every time on these i.e. there's multiple occurrences on a single line?
- would be nice to split this over multiple lines

MikeMcQuaid replied:
Also: would be good to handle any naming format for the second directory, ideally and not just `homebrew-`. Alternatively: remove directories with a different format.

metacollin replied:
Ah, good call.  That used to be necessary when the string wasn't split into lines but now I don't think it matters.

MikeMcQuaid commented on Library/Homebrew/cmd/formulae.sh:
```diff
@@ -0,0 +1,11 @@
+homebrew-formulae() {
+  local formulae="$(find "$HOMEBREW_REPOSITORY"/Library/Taps -type d \( -name Casks -o -name cmd -o -name .github \) -prune -false -o -name '*rb' | sed 's/\.rb//g' | sed -E 's .*/Taps/(.*)/homebrew- \1/ g' | sed 's /Formula/ / g')"
+  local shortnames="$(echo "$formulae" | cut -d / -f 3)"
+  echo -e "$formulae \n $shortnames" | grep -v '^homebrew/' | sort -uf
```
Could the `grep-v` instead be part of one of the previous lines?

metacollin replied:
I don't think so.  All the formula are only listed by their full name according to the tap folder structure (so, homebrew/core/<formula names>).  If those lines are removed before the `cut -d / -f 3` command (which removes everything before the final `/`), then none of the homebrew core formula will get included.  

I actually thought about doing it this way for a while, but this is by far the most efficient method I could come up with.    By removing the extended name of the core homebrew formula in the list, then those names can be generated with the short names of everything else and simply have those longer names omitted at the end (`grep -v`).  Otherwise, an entire separate 'pipeline' so to speak is needed just for the special case of the homebrew core names.

MikeMcQuaid replied:
Ok, makes sense 👍🏻

MikeMcQuaid commented on completions/zsh/_brew:
```diff
@@ -111,6 +111,7 @@ __brew_common_commands() {
     'edit:edit a formula'
     'fetch:download formula resources to the cache'
     'formula:the path for a formula'
+		'formulas:show a sorted list of installable formulas'
```
```suggestion
		'formulae:show a sorted list of installable formulae'
```

MikeMcQuaid reviewed (commented):
Nice work so far here!

MikeMcQuaid commented on Library/Homebrew/cmd/formulae.sh:
```diff
@@ -0,0 +1,11 @@
+#:  * `formulae`
+#:
+#:  Prints a sorted list of locally available formulas including shortnames.
```
What is a `shortname`? that's not a term we use elsewhere.

metacollin replied:
Shortname was originally used in homebrew's bash completion:

```bash
# completions/bash/brew:53 

__brew_complete_formulae() {
  local cur="${COMP_WORDS[COMP_CWORD]}"
  local formulas="$(brew search)"
  local shortnames="$(echo "$formulas" | \grep / | \cut -d / -f 3)"
  COMPREPLY=($(compgen -W "$formulas $shortnames" -- "$cur"))
}
```

For completion, we need both the full formula path ( `repo_owner/tap_name/formula_name`) as well as simply `formula_name`.  `shortnames` is just a temporary holder for the formula names from taps without the rest of the path.  That does mean there will be duplicated names, though every line is still unique.  

Depending on what you think the behavior should be, we could either:

1.  Leave it as is
2.  Add an argument (`--include-shortened-names` or sometihng similar.  I am not sure what the brew terminology is or if it even exists) that will include those extra names for completions, but by default there will be only the full names
3. Omit it entirely and move this logic into the completions themselves.  
4. Do something else entirely

I have no preference personally, and I don't think it would impact performance all that much.

MikeMcQuaid replied:
```suggestion
#:  Prints a sorted list of locally available formulae.
```

I would just go for this instead, I think it's fine 👍🏻 

Are you saying this should include both:
```
wget
some/tap/wget
``` 

in the output. If so could be:

```suggestion
#:  Prints a sorted list of available formulas names with and without tap prefixes.
```

MikeMcQuaid commented on Library/Homebrew/cmd/formulae.sh:
```diff
@@ -0,0 +1,11 @@
+#:  Prints a sorted list of locally available formulas including shortnames.
+#:
+
+# shellcheck disable=SC2155
```
What's this for? Would be good to remove it.

metacollin replied:
Yeah sorry looks like I needed to check everything more carefully before the commit.  Thank you catching/cleaning up a lot of it.

metacollin replied:
`shellcheck: warning
warning - Declare and assign separately to avoid masking return values. [SC2155]`

It gets triggered any time you try to declare and assign a local variable in the same line.  It doesn't matter if one isn't using the return value of the command inside the `"$()"`.  

The alternative to disabling that check is this:

```bash
homebrew-formulae() {
  local formulae
  formulae="$(find "$HOMEBREW_REPOSITORY"/Library/Taps -type d \( -name Casks -o -name cmd -o -name .github \) -prune -false -o -name '*rb' | sed 's/\.rb//g' | sed -E 's .*/Taps/(.*)/homebrew- \1/ g'>
  local shortnames
  shortnames="$(echo "$formulae" | cut -d / -f 3)"
  echo -e "$formulae \n $shortnames" | grep -v '^homebrew/' | sort -uf
}```

which also makes the command slightly slower.  

You can see some examples with `local` on the shellcheck [wiki](https://github.com/koalaman/shellcheck/wiki/SC2155)

MikeMcQuaid replied:
> which also makes the command slightly slower, and adds two more fairly pointless lines IMO.

Are you sure it's slower? How much slower?

Yeh, I think the separate declarations is preferable 👍🏻

metacollin replied:
Well, I can't say how or why, but there was a definite speed impact how it was originally written... but the newer version with several changes based on everyone's suggestions seems to have fixed that speed hit.  Or maybe it was just a fluke?  I don't really know :(.  Regardless, I've made the suggested changes.

maxim-belkin commented on Library/Homebrew/cmd/formulae.sh:
```diff
@@ -0,0 +1,13 @@
+# shellcheck disable=SC2155
+homebrew-formulae() {
+  local formulae="$(
+    find "$HOMEBREW_REPOSITORY"/Library/Taps -type d \( -name Casks -o -name cmd -o -name .github \) -prune -false -o -name '*\.rb' 
```
Line continuation?
```suggestion
    find "$HOMEBREW_REPOSITORY"/Library/Taps -type d \( -name Casks -o -name cmd -o -name .github \) -prune -false -o -name '*\.rb' \
```

maxim-belkin commented on Library/Homebrew/cmd/formulae.sh:
```diff
@@ -0,0 +1,13 @@
+homebrew-formulae() {
+  local formulae="$(
+    find "$HOMEBREW_REPOSITORY"/Library/Taps -type d \( -name Casks -o -name cmd -o -name .github \) -prune -false -o -name '*\.rb' 
+      | sed -E -e 's/\.rb//g' -e 's .*/Taps/(.*)/homebrew-\1/ g' -e 's /Formula/ / g')"
```
do we need `-E` here?
`'s .*/Taps/(.*)/homebrew-\1/ g'` seems malformed. Did you mean `-e 's|.*/Taps/\(.*\)|homebrew-\1|g'`? (`\(` and `\)` instead of `(` and `)` without `-E`.

I'd suggest to `|` symbol for splitting patterns because it let's us avoid escaping `/` symbol in the second substitution.

metacollin replied:
Can't use that commit as-is, or it will fail one of shellcheck's tests.  

Apparently it has strong opinions on where the `|` goes.  Sorry, I didn't see that you had made a suggestion before I committed the changes, but its essentially your commit but with the | moved so shellcheck doesn't complain.  

@MikeMcQuaid I also added a couple more filters to the find command, since the recent inclusions of the new homebrew-bundler and several other new folders inside the homebrew/homebrew-core tap folder were getting picked up by the search.  Fortunately, adding directory names to skip has virtually no impact on the speed.  I wish I could limit the search to the Formula directory, but that would break backwards compatibility with older taps that just have the formula in the tap folder.

metacollin commented:
Thanks everyone for your suggestions and help.  Keep the opinions/concerns coming and I'll try to get this to a point that everyone is happy with it!

MikeMcQuaid reviewed (commented):
This is looking great! We're almost there and I'm excited by the speedup!

maxim-belkin commented on docs/Manpage.md:
```diff
@@ -233,6 +233,10 @@ For tarballs, also print SHA-256 checksums.
 
+### `formulae`
+
+Prints a sorted list of locally available formulas including shortnames.
```
1. Use the root form of the verb "print" -- "Print a sorted list..."
2. Homebrew uses British English, so... formulae.
3. I'm not sure "shortname" is a word. Did you mean "... including their short names"?

MikeMcQuaid replied:
I'd say "output" instead of "print". Agreed on short name (have made that point above, I think).

EricFromCanada replied:
In my writing I've been using "output" only as a noun, preferring either "show" (the default), "display" (for single line output), "list" (for lists), "print" or "include" (for things that only slightly change or append to the output). So in this case, I'd write "List all locally available formulae including short names."

maxim-belkin commented on completions/zsh/_brew:
```diff
@@ -111,6 +111,7 @@ __brew_common_commands() {
     'edit:edit a formula'
     'fetch:download formula resources to the cache'
     'formula:the path for a formula'
+		'formulae:show a sorted list of installable formulae'
```
Perhaps, change installable to available (?).

MikeMcQuaid replied:
Sounds good 👍🏻

metacollin replied:
Changed!

metacollin commented on Library/Homebrew/cmd/formulae.sh:
```diff
@@ -0,0 +1,13 @@
+homebrew-formulae() {
+  local formulae="$(
+    find "$HOMEBREW_REPOSITORY"/Library/Taps -type d \( -name Casks -o -name spec -o -name lib -o -name cmd -o -name .github \) -prune -false -o -name '*\.rb' |
+    sed -E -e 's/\.rb//' -e 's .*/Taps/(.*)/homebrew- \1/ ' -e 's /Formula/ / ')"
```
You're right, I've updated it.  However, this won't work as is, because the '|' symbol is reserved for boolean or, so I changed it to underscore (unless anyone has a better idea).

metacollin reviewed (commented):
OK, sorry I took almost a month to get around to this.

metacollin reviewed (commented):
👍

MikeMcQuaid commented on Library/Homebrew/brew.sh:
```diff
@@ -38,12 +38,13 @@ HOMEBREW_TEMP="${HOMEBREW_TEMP:-${HOMEBREW_DEFAULT_TEMP}}"
-  --caskroom)          echo "$HOMEBREW_PREFIX/Caskroom"; exit 0 ;;
-  --cache)             echo "$HOMEBREW_CACHE"; exit 0 ;;
-  shellenv)            source "$HOMEBREW_LIBRARY/Homebrew/cmd/shellenv.sh"; homebrew-shellenv; exit 0 ;;
+  --prefix)             echo "$HOMEBREW_PREFIX"; exit 0 ;;
+  --cellar)             echo "$HOMEBREW_CELLAR"; exit 0 ;;
+  --repository|--repo)  echo "$HOMEBREW_REPOSITORY"; exit 0 ;;
+  --caskroom)           echo "$HOMEBREW_PREFIX/Caskroom"; exit 0 ;;
+  --cache)              echo "$HOMEBREW_CACHE"; exit 0 ;;
+  shellenv)             source "$HOMEBREW_LIBRARY/Homebrew/cmd/shellenv.sh"; homebrew-shellenv; exit 0 ;;
+  formulae)    source "$HOMEBREW_LIBRARY/Homebrew/cmd/formulae.sh"; homebrew-formulae; exit 0 ;;
```
```suggestion
  --prefix)            echo "$HOMEBREW_PREFIX"; exit 0 ;;
  --cellar)            echo "$HOMEBREW_CELLAR"; exit 0 ;;
  --repository|--repo) echo "$HOMEBREW_REPOSITORY"; exit 0 ;;
  --caskroom)          echo "$HOMEBREW_PREFIX/Caskroom"; exit 0 ;;
  --cache)             echo "$HOMEBREW_CACHE"; exit 0 ;;
  shellenv)            source "$HOMEBREW_LIBRARY/Homebrew/cmd/shellenv.sh"; homebrew-shellenv; exit 0 ;;
  formulae)            source "$HOMEBREW_LIBRARY/Homebrew/cmd/formulae.sh"; homebrew-formulae; exit 0 ;;
```

maxim-belkin commented on Library/Homebrew/cmd/formulae.sh:
```diff
@@ -0,0 +1,28 @@
+
+homebrew-formulae() {
+  local formulae
+  formulae="$( \
```
```suggestion
  formulae="$(
```

MikeMcQuaid commented:
Thanks so much for your great, first contribution! Without people like you submitting PRs we couldn't run this project. You rock, @metacollin!
